### ISSUE TITLE
ignore pattern exclusion causes CPU hog with large subtrees

### ISSUE THREAD

**Issue by cptMikky** [syncthing/syncthing#7860]
Ignore pattern with exclusion causes the subtree to be recursively walked in full. That wouldn't be an issue if the walk didn't hog 8-core cpu even when scheduled in the absolute lowest priority possible (SCHED_IDLEPRIO - see MuQSS for details).

The pattern is
```
!**/.git/config
/git
```
The use-case is obvious - do not sync anything except git configuration.

There's about 175 000 files under `/git`. While walking through that subtree, syncthing spawns 100+ threads in under an hour, hogs CPU and is able to process about 1-5 files per second (watching live logs with STTRACE=fs,walkfs,events).

Linux 5.12.14 with MuQSS CPU scheduler (NixOS unstable, encrypted ZFS dataset, NVME disk, RAM usage ~1G)
Syncthing v1.18.0

Note that with upstream kernel with CFS scheduler the result is even worse - with MuQSS the system is _barely_ responsive (i.e. I can switch to a console and kill syncthing - this takes about 1-2 minutes), with CFS (SCHED_IDLE) in many cases the system becomes unresponsive to the point where manual shutdown is needed.

Also the GOMAXPROCS env var is apparently ignored (same results regardless of it's value).

**Comment by AudriusButkevicius** [comment]
There is no immediate resolution to this.

I think you should start a thread on the forum first to discuss the possible options, if there are any (not obvious to me).

**Comment by cptMikky** [comment]
@AudriusButkevicius what about the GOMAXPROCS issue? From what I've read around at least this should keep syncthing down on SMP machines... The only thing I really care about is system responsiveness.

**Comment by AudriusButkevicius** [comment]
Go spawns a thread per io operation, as there is no such thing as non-blocking IO on certain platforms, GOMAXPROCS has nothing to do with that.

**Comment by cptMikky** [comment]
Then I suggest at least clearing that up in the docs/faq as this is often recommended as a "solution" to common "cpu-hog" issues but from what I understand this cannot possibly help with most of them as they would be IO-bound.

What I do not understand is why a thread blocked on IO causes high CPU usage (I've also scheduled syncthing with IDLE IO prio but it seems that scheduling in general doesn't have much effect on this specific issue. on Linux that is). Are you saying that this is a bug in GO?

**Comment by AudriusButkevicius** [comment]
As I suggested, this is not the best place to discuss this.

There is not enough information to understand what exactly is happening.

In theory, a spawned thread blocked on IO should not cause CPU usage, but it might vary how the underlying filesystem behaves. Perhaps it's the context switching which causes CPU load.
Perhaps it's something else, I have no idea what state syncthing is in when this happens, so I can't say much.

There are knobs that can be tweaked, but as I suggested, this is probably best done in the forum.
